vlm/src/main/scala/LayoutTileSource.scala

```diff
@@ -0,0 +1,87 @@
+  *
+  * @param source raster source that can be queried by bounding box
+  * @param layout definition of a tile grid over the pixel grid
+  */
+class LayoutTileSource(val source: RasterSource, val layout: LayoutDefinition) {
+  // TODO: require that source and layout are grid alligned, consider floating point precision
+  def sourceColOffset: Long = ((source.extent.xmin - layout.extent.xmin) / layout.cellwidth).toLong
+  def sourceRowOffset: Long = ((layout.extent.ymax - source.extent.ymax) / layout.cellheight).toLong
+
+  /** Read tile according to key.
+    * If tile area intersects source partially the non-intersecting pixels will be filled with NODATA.
+    * If tile area does not intersect source None will be returned.
+    */
+  def read(key: SpatialKey): Option[MultibandTile] = {
+    val col = key.col.toLong
+    val row = key.row.toLong
+    val sourcePixelBounds = GridBounds(
+        colMin = (col * layout.tileCols - sourceColOffset).toInt,
+        rowMin = (row * layout.tileRows - sourceRowOffset).toInt,
+        colMax = ((col+1) * layout.tileCols - 1 - sourceColOffset).toInt,
+        rowMax = ((row+1) * layout.tileRows - 1 - sourceRowOffset).toInt)
+    for {
+      bounds <- sourcePixelBounds.intersection(source)
+      raster <- source.read(bounds)
+    } yield {
+      // raster is smaller but not bigger than I think ...
+      // its offset is relative to the raster we wished we had
+      val colOffset = bounds.colMin - sourcePixelBounds.colMin
+      val rowOffset = bounds.rowMin - sourcePixelBounds.rowMin
+      raster.tile.mapBands{ (_, band) =>
+        PaddedTile(band, colOffset, rowOffset, layout.tileCols, layout.tileRows)
+      }
+    }
+  }
+
+  /** Read multiple tiles according to key.
+    * If each tile area intersects source partially the non-intersecting pixels will be filled with NODATA.
+    * If tile area does not intersect source it will be excluded from result iterator.
+    */
+  def readAll(keys: Iterator[SpatialKey]): Iterator[(SpatialKey, MultibandTile)] = {
+    for {
+      key <- keys
+      col = key.col.toLong
+      row = key.row.toLong
+      sourcePixelBounds = GridBounds(
+        colMin = (col * layout.tileCols - sourceColOffset).toInt,
+        rowMin = (row * layout.tileRows - sourceRowOffset).toInt,
+        colMax = ((col+1) * layout.tileCols - 1 - sourceColOffset).toInt,
+        rowMax = ((row+1) * layout.tileRows - 1 - sourceRowOffset).toInt)
+      bounds <- sourcePixelBounds.intersection(source)
+      raster <- source.read(bounds)
+    } yield {
+      // raster is smaller but not bigger than I think ...
+      // its offset is relative to the raster we wished we had
+      val colOffset = bounds.colMin - sourcePixelBounds.colMin
+      val rowOffset = bounds.rowMin - sourcePixelBounds.rowMin
+      val tile = raster.tile.mapBands{ (_, band) =>
+        PaddedTile(band, colOffset, rowOffset, layout.tileCols, layout.tileRows)
+      }
+      (key, tile)
+    }
+  }
```

Comment:
Going off my earlier comment, if we knew what keys intersected the `RasterSource` we could have a `readIntersectingKeys` method where the keys that are being read are the ones which intersect the `RasterSource`.